src/nvme/util.c

```diff
@@ -141,3 +142,258 @@ int nvme_admin(struct nvme_ctrl *ctrl, union nvme_cmd *sqe, void *buf, size_t le
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static inline int __map_prp_first(leint64_t *prp1, leint64_t *prplist, uint64_t iova, size_t len,
+			      int pageshift)
+{
+	size_t pagesize = 1 << pageshift;
+	int max_prps = 1 << (pageshift - 3);
+
+	/* number of prps required to map the buffer */
+	int prpcount = 1;
+
+	*prp1 = cpu_to_le64(iova);
+
+	/* account for what is covered with the first prp */
+	len -= min_t(size_t, len, pagesize - (iova & (pagesize - 1)));
+
+	/* any residual just adds more prps */
+	if (len)
+		prpcount += (int)ALIGN_UP(len, pagesize) >> pageshift;
+
+	if (prpcount > 1 && !ALIGNED(iova, pagesize))
+		/* align down to simplify loop below */
+		iova = ALIGN_DOWN(iova, pagesize);
+
+	if (prpcount > max_prps) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	/*
+	 * Map the remaining parts of the buffer into prp2/prplist. iova will be
+	 * aligned from the above, which simplifies this.
+	 */
+	for (int i = 1; i < prpcount; i++)
+		prplist[i - 1] = cpu_to_le64(iova + ((uint64_t)i << pageshift));
+
+	/*
+	 * prpcount may be zero if the buffer length was less than the page
+	 * size, so clamp it to 1 in that case.
+	 */
+	return clamp_t(int, prpcount, 1, prpcount);
+}
+
+static inline int __map_prp_append(leint64_t *prplist, uint64_t iova, size_t len, int max_prps,
+				   int pageshift)
+{
+	int prpcount = max_t(int, 1, (int)len >> pageshift);
+	size_t pagesize = 1 << pageshift;
+
+	if (prpcount > max_prps) {
+		log_error("too many prps required\n");
+
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (!ALIGNED(iova, pagesize)) {
+		log_error("unaligned iova 0x%" PRIx64 "\n", iova);
+
+		errno = EINVAL;
+		return -1;
+	}
+
+	for (int i = 0; i < prpcount; i++)
+		prplist[i] = cpu_to_le64(iova + ((uint64_t)i << pageshift));
+
+	return prpcount;
+}
+
+static inline void __set_prp2(leint64_t *prp2, leint64_t prplist, leint64_t prplist0, int prpcount)
+{
+	if (prpcount == 2)
+		*prp2 = prplist0;
+	else if (prpcount > 2)
+		*prp2 = prplist;
+	else
+		*prp2 = 0x0;
+}
+
+int nvme_map_prp(struct nvme_ctrl *ctrl, leint64_t *prplist, uint64_t prplist_iova,
+		 union nvme_cmd *cmd, uint64_t iova, size_t len)
+{
+	int prpcount;
+	int pageshift = __mps_to_pageshift(ctrl->config.mps);
+
+	prpcount = __map_prp_first(&cmd->dptr.prp1, prplist, iova, len, pageshift);
+	if (prpcount < 0) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	__set_prp2(&cmd->dptr.prp2, cpu_to_le64(prplist_iova), prplist[0], prpcount);
+
+	return 0;
+}
+
+int nvme_mapv_prp(struct nvme_ctrl *ctrl, leint64_t *prplist,
+		  union nvme_cmd *cmd, struct iovec *iov, int niov)
+{
+	struct iommu_ctx *ctx = __iommu_ctx(ctrl);
+
+	size_t len = iov->iov_len;
+	int pageshift = __mps_to_pageshift(ctrl->config.mps);
+	size_t pagesize = 1 << pageshift;
+	int max_prps = 1 << (pageshift - 3);
+	int ret, prpcount;
+	uint64_t iova, prplist_iova;
+
+	if (nvme_translate_vaddr(ctrl, iov->iov_base, &iova))
+		return -1;
+
+	if (nvme_translate_vaddr(ctrl, prplist, &prplist_iova))
+		return -1;
```

Comment:
following on what i mentioned above, here you've added a helper that first checks the iommu, then the cmb. But, we can simplify and just always have it in the iommu.